xtrlock.c

```diff
@@ -166,48 +205,103 @@ int main(int argc, char **argv){
+            if (ev.xkey.time < timeout) { XBell(display,0); break; }
+            clen= XLookupString(&ev.xkey,cbuf,9,&ks,0);
+            switch (ks) {
+                case XK_Escape: 
+                case XK_Clear:
+                    rlen=0; break;
+                case XK_Delete: 
+                case XK_BackSpace:
+                    if (rlen>0) rlen--;
+                    break;
+                case XK_Linefeed: 
+                case XK_Return:
+                    if (rlen==0) break;
+                    else rbuf[rlen]=0;
+                    if (passwordok(rbuf)) goto loop_x;
+                    XBell(display,0);
+                    rlen= 0;
+                    if (timeout) {
+                        goodwill+= ev.xkey.time - timeout;
+                        if (goodwill > MAXGOODWILL) {
+                            goodwill= MAXGOODWILL;
+                        }
+                    }
+                    timeout= -goodwill*GOODWILLPORTION;
+                    goodwill+= timeout;
+                    timeout+= ev.xkey.time + TIMEOUTPERATTEMPT;
+                    break;
+                default:
+                if (clen != 1) break;
+                /* allow space for the trailing \0 */
+	            if (rlen < (sizeof(rbuf) - 1)){
+	                rbuf[rlen]=cbuf[0];
+	                rlen++;
+	            }
+                break;
+            }break;
+
+        default: break;
     }
-  }
- loop_x:
+  }/*end checker loop*/
+ loop_x:/*loop exit*/
   exit(0);
 }
+
+int main(int argc, char **argv){/*TODO:get rid of root access when not necessary*/
+    bool need_lock = false;
+        /*area for any arg init*/
+    if(!init_cust_pw
+    ){  fprintf(stderr,"Failed to init custom password config");
+        exit(-1);}
+        /*area for any arg init*/
+
+        char opt = 0;/*TODO: fix changes*/
+        while((opt = getopt(argc, argv, "h:p:e:c:l")) != -1){
+
+                if('h' == opt){/*help*/
+                    print_help();
+                    exit(0);
+                }
+                if('p' == opt){/*custom pwd without encryption*/
+                    if(strlen(optarg) <= 1){
+                            fprintf(stderr, "One character password is disallowed\n");
+                            return false;
+                    }
+                    cust_pw_setting.enable = true;
+                    //cust_pw_setting.pwd = crypt(argv[2], argv[2]);
+                    cust_pw_setting.pwd = strdup(crypt(optarg, optarg));/*never freed, fine in this case*/
+                    cust_pw_setting.crypt = false;
+                    need_lock = true;
+                    for(int i=0; i<100; i++){
+                            if(NULL == cust_pw_setting.pwd){
+                                fprintf(stderr,"strdup:%s\n", strerror(errno)); 
+                                cust_pw_setting.pwd = strdup(crypt(optarg, optarg));/*never freed, fine in this case*/
+                            }
+                    }
```

Comment:
Nooooooooooo!!!1!

Brief primer: If ```strdup``` fails and returns ```NULL``` that means you're out of memory. That condition can persist for several seconds or even minutes. Dumbly retrying a certain number of times does not work!

If ```strdup``` returns ```NULL```, the only sane way of action is to *abort operation right now with a visible error message*. Now, that would be a problem if the screen was already locked, but in this case it is perfectly fine as the screen locker is only about to start.

It may make sense to flash the screen briefly on successful startup to give the user an easily visible notification that her screen has *indeed* been locked.

BTW, what do you mean with the "never freed[...]" comment?